backend/core/src/csv/csv-builder.service.ts

```diff
@@ -1,130 +1,330 @@
+        })
       } else {
-        headers.push(metadata.label)
+        for (let i = 0, keys = Object.keys(val); i < keys.length; i++) {
+          this.parseApplication(
+            keys[i],
+            val[keys[i]],
+            header.length ? `${header} ${keys[i]}` : keys[i],
+            index
+          )
+        }
       }
+    } else {
+      const [newKey, mappedField] = this.setHeader(header, index)
+      // format data before pushing
+      if (typeof val === "boolean") {
+        val = val ? "Yes" : "No"
+      }
+      // use JSON.stringify to escape double and single quotes
+      this.data.push({
+        header: newKey,
+        mappedField,
+        val: val !== undefined && val !== null ? JSON.stringify(val) : "",
+      })
     }
-    return headers
   }
 
-  private normalizeMetadataArrays(
-    arr: any[],
-    formattingMetadataAggregate: FormattingMetadataAggregate
-  ) {
-    formattingMetadataAggregate
-      .filter((metadata) => metadata.type === "array")
-      .forEach((metadata) => {
-        const md = metadata as FormattingMetadataArray
-        if (md.size !== null) {
-          return
-        }
-        md.size = Math.max(
-          ...arr.map((item) => {
-            const value = this.retrieveValueByDiscriminator(item, md.discriminator)
-            if (!value || !Array.isArray(value)) {
-              return 0
+  public build(arr: any[], includeDemographics?: boolean): string {
+    if (this.inProgress === true) {
+      throw new HttpException(
+        {
+          status: HttpStatus.TOO_MANY_REQUESTS,
+          error: "Download Already in Progress",
+        },
+        HttpStatus.TOO_MANY_REQUESTS
+      )
+    }
+    this.setProgress(true)
+    let dataString = ""
+    try {
+      if (!includeDemographics) {
+        this.addToExlucdedKeys("demographics")
+      }
+      arr.forEach((row: ApplicationDto) => {
+        this.parseApplication("", row, "")
+        this.data.push({ header: "_endRow", mappedField: undefined, val: "_endRow" })
+      })
+
+      // filter out excluded keys and fields not in mappedFields
+      let headerArray = Object.keys(this.headers).filter((header) => {
+        const mappedField =
+          this.mappedFields[header.replace(HEADER_COUNT_REG, "")] ||
+          this.mappedFields[header.split(" ")[0]]
+        return !this.getExcludeKeysRegex().test(header) && mappedField
+      })
+
+      // initiate arrays to insert data
+      const rows = Array.from({ length: arr.length }, () => Array(headerArray.length))
+
+      headerArray = headerArray
+        .map((column) => {
+          let columnString = ""
+          const test1 = this.mappedFields[column.replace(HEADER_COUNT_REG, "")]
+          const test2 = this.mappedFields[column.split(" ")[0]]
+          let mappedField = test1 ?? test2
+          if (mappedField) {
+            if (!test1 && test2) {
+              const [, ...fields] = column.split(" ")
+              fields.forEach((field) => {
```

Comment:
Are fields actually tokens of a header? Maybe a better name would be `headerTokens`